consensus/hotstuff/eventhandler/event_handler_test.go

```diff
@@ -388,29 +307,160 @@ func (es *EventHandlerSuite) SetupTest() {
+// should not vote.
+func (es *EventHandlerSuite) TestOnReceiveProposal_NoVote() {
+	proposal := createProposal(es.initView, es.initView-1)
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+
+	// I'm the next leader
+	es.committee.leaders[es.initView+1] = struct{}{}
+	// no vote for this proposal
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+	es.voteAggregator.AssertCalled(es.T(), "AddBlock", proposal)
+}
+
+// TestOnReceiveProposal_NoVote_ParentProposalNotFound tests scenario: received a valid proposal for cur view, no parent for this proposal found
+// should not vote.
+func (es *EventHandlerSuite) TestOnReceiveProposal_NoVote_ParentProposalNotFound() {
+	proposal := createProposal(es.initView, es.initView-1)
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+
+	// remove parent from known proposals
+	delete(es.forks.proposals, proposal.Block.QC.BlockID)
+
+	// no vote for this proposal, no parent found
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+	es.voteAggregator.AssertCalled(es.T(), "AddBlock", proposal)
+}
+
+// TestOnReceiveProposal_Vote_NextLeader tests scenario: received a valid proposal for cur view, safe to vote, I'm the next leader
+// should vote and add vote to VoteAggregator.
+func (es *EventHandlerSuite) TestOnReceiveProposal_Vote_NextLeader() {
+	proposal := createProposal(es.initView, es.initView-1)
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+	es.voteAggregator.On("AddVote", mock.Anything).Return().Once()
+
+	// I'm the next leader
+	es.committee.leaders[es.initView+1] = struct{}{}
+
+	// proposal is safe to vote
+	es.safetyRules.votable[proposal.Block.BlockID] = struct{}{}
+
+	// vote should be created for this proposal
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+}
+
+// TestOnReceiveProposal_Vote_NextLeader tests scenario: received a valid proposal for cur view, safe to vote, I'm not the next leader
+// should vote and send vote to next leader.
+func (es *EventHandlerSuite) TestOnReceiveProposal_Vote_NotNextLeader() {
+	proposal := createProposal(es.initView, es.initView-1)
+	es.voteAggregator.On("AddBlock", proposal).Return(nil).Once()
+
+	// proposal is safe to vote
+	es.safetyRules.votable[proposal.Block.BlockID] = struct{}{}
+
+	// vote should be created for this proposal
+	err := es.eventhandler.OnReceiveProposal(proposal)
+	require.NoError(es.T(), err)
+	require.Equal(es.T(), es.endView, es.paceMaker.CurView(), "incorrect view change")
+
+	lastCall := es.communicator.Calls[len(es.communicator.Calls)-1]
+	// the last call is SendVote
+	require.Equal(es.T(), "SendVote", lastCall.Method)
```

Comment:
Why are you doing it this way, rather than just `AssertCalled("SendVote", ...)`?

Reply:
Basically to avoid creating vote to match with, I guess it gives a little bit more flexibility.